Add AI agents authentication deep dive (#3284)
giantswarm/docs: src/content/overview/ai-agents/authentication/index.md

---
title: 'Authentication deep dive: from your MCP client to the Kubernetes API'

> linkTitle: Authentication deep dive

Reviewer: I'd prefer a shorter handle in the menu here

<img width="270" height="281" alt="Image" src="https://github.com/user-attachments/assets/bdbd0979-cce7-4a93-88a0-e0db68dc9662" />

```suggestion
linkTitle: Authentication
```

Reviewer 2: Should the title not be agent platform instead or AI Agent?

Reviewer 2: it makes me think fleet management is now some kind of misdirection right? It is a fleet of what ?

Author: I changed the handle in the menu. 

> Should the title not be agent platform instead or AI Agent?

Renaming it would touch every page in the section plus its menu identifier. I'd rather not couple that to this page; happy to discuss it in a separate issue/PR

Reviewer 2: Oh you're totally right. I think this is more of a question to @sig-docs

Author: @QuentinBisson that's a fair question, but it's about the parent section (`overview/ai-agents`), which predates this PR — renaming it would touch every page in the section plus its menu identifier. I'd rather not couple that to this page; happy to discuss it in a separate issue/PR if you want to push for "Agent platform".

MCP servers will respond back to Claude Code with the information. And the ones that require authentication will respond with a `401 Unauthorized` HTTP response. This lets Claude Code know that it needs to send authenticated requests, meaning that the user needs to log in. That's why when you run the `/mcp` command in Claude Code, you'll see the "needs authentication" message next to an MCP server. To log in, you just need to select the MCP server from the list. Doing it will start the OAuth authentication flow.

![Claude Code's /mcp view showing MCP servers that need authentication](claude-mcp-needs-auth.png)

Muster is itself the Authorization Server for MCP clients like Claude Code. Behind the scenes, it delegates the actual user authentication to Dex. Muster is registered as a client of Dex, which lets us plug in different identity providers (Dex connectors) like Google or Azure without changing the architecture.

## What happens when you start the OAuth flow to log in to Muster

### 1. Claude Code → Muster

> When Claude Code gets a `401 Unauthorized` HTTP response, the response carries a `WWW-Authenticate` header pointing at the server's metadata, which is how Claude Code discovers the authorization server and its endpoints. Using that information, Claude Code (the client) opens your browser at Muster's `/oauth/authorize`, identifying itself with its `client_id`. For MCP clients that `client_id` isn't a pre-registered value but a URL the client controls (see ["How Claude Code becomes a client without pre-registration"](#how-claude-code-becomes-a-client-without-pre-registration) below).

Reviewer 3: ```suggestion
When Claude Code gets a `401 Unauthorized` HTTP response, the response carries a `WWW-Authenticate` header pointing at the server's metadata, which is how Claude Code discovers the Authorization server (AS) and its endpoints. Using that information, Claude Code (the client) opens your browser at Muster's `/oauth/authorize`, identifying itself with its `client_id`. For MCP clients, that `client_id` isn't a pre-registered value but a URL the client controls (see ["How Claude Code becomes a client without pre-registration"](#how-claude-code-becomes-a-client-without-pre-registration) below).
```

### 2. Muster → Dex

Muster doesn't know your password and doesn't want to. Instead of showing a login form, it redirects that same browser onward to Dex's `/auth` endpoint. Now Muster acts as Dex's client, using the client_id/client_secret that was pre-registered for Muster in Dex's configuration (a `staticClients` entry in Dex's config. Muster reads its copy from its own config).

### 3. Dex → the upstream identity provider

Dex is doing the exact same trick one level deeper: it's an AS to Muster but a client of the identity provider, for example Google. Each Dex "connector" holds a client_id/secret registered with the upstream identity provider. It redirects the browser to the identity provider's login page.

### 4. User logs in

You type your password at the identity provider login page, and only there. Neither Dex, nor Muster, nor Claude Code ever sees it.

### 5. The chain unwinds

> The identity provider redirects the browser back to Dex's callback with a code. Dex exchanges it (authenticating with its identity provider / connector `client_secret`), mints its own tokens, and redirects the browser to Muster's `/oauth/callback` with a new code. Muster exchanges that code at Dex's `/token` endpoint, authenticating with Muster's Dex `client_secret`, and receives the Dex ID token. Muster then redirects the browser one last time to Claude Code's localhost callback with yet another code. When Claude Code exchanges that code, Muster mints and returns its own access token.

Reviewer 3: IMO it is a bit confusing, explaning why exchange the cods may help users to understand why so many tokens are involved

Author: Added a small paragraph trying to explain it better.

> The identity provider redirects the browser back to Dex's callback with a code. Dex exchanges it (authenticating with its identity provider / connector `client_secret`), mints its own tokens, and redirects the browser to Muster's `/oauth/callback` with a new code. Muster exchanges that code at Dex's `/token` endpoint, authenticating with Muster's Dex `client_secret`, and receives the Dex ID token. Muster then redirects the browser one last time to Claude Code's localhost callback with yet another code. When Claude Code exchanges that code, Muster mints and returns its own access token.

Reviewer: This also looks like a good case for step/steps (see my other comment)

Author: done

### 3. Does Dex vouch for the user right now

On every request coming from MCP clients, Muster will call the `/userinfo` endpoint in Dex to check if the Dex access token is still valid at the Dex level. This catches revocation at Dex immediately (session deleted or revoked at Dex). Note that this check only proves the session is still alive at Dex. A user removed at the upstream identity provider (Google / Azure) isn't visible to `/userinfo`. Dex itself only finds out the next time it refreshes against the upstream, which is [check 2 described earlier](#2-is-the-underlying-dex-session-still-alive). Each layer vouches for itself in real time. News from the upstream layer arrives at refresh time.

### 4. Was this token minted for this server

> When it mints a token, Muster records in its storage which resource the token was requested for (its own resource URI—this is OAuth's "resource indicator," RFC 8707). At validation time it checks that binding, so a token minted for a different resource server is rejected even if it's otherwise valid.

Reviewer: For such step-by-step content we have the HUGO shortcode `steps` and `step`. Would be nice to use it here. You'll find examples in the code base.

Author: done, ptal

> When it mints a token, Muster records in its storage which resource the token was requested for (its own resource URI—this is OAuth's "resource indicator," RFC 8707). At validation time it checks that binding, so a token minted for a different resource server is rejected even if it's otherwise valid.

Reviewer 3: ```suggestion
When Dex mints a token, Muster records in its storage which resource the token was requested for (its own resource URI—this is OAuth's "resource indicator," RFC 8707). At validation time, it checks that binding, so a token minted for a different resource server is rejected even if it's otherwise valid.
```

Author: Actually, this check is about Muster's own opaque access token, which Muster mints and stores (Dex never sees it). I replaced the pronoun for the explicit name.

Author: Good catch on the ambiguous pronoun — but I went the other direction: this check is about Muster's own opaque access token, which Muster mints and stores (Dex never sees it). The RFC 8707 resource binding lives in Muster's token storage, which is why the check works at all. I've reworded it to "When Muster mints an access token, it records in its storage…" so the subject is explicit. Thanks for flagging it!

But in our scenario, `muster` is the one requesting tokens from Dex. If the token only contains `audience: muster`, Kubernetes would reject the ID token due to having the wrong audience. To work around this, when `muster` requests tokens from Dex it requests to include the audience that Kubernetes expects in the audience field of the token. Dex then mints the token with both audiences.

Can any client just request tokens addressed to any other client? No, this is Dex's cross-client audience feature, and it's deny-by-default. The target client's Dex config must explicitly list the requester in its `trustedPeers` config.

Similarly, `mcp-kubernetes` is configured to accept tokens that contain `muster` in the audience field, via the `trustedAudiences` list in `mcp-kubernetes` configuration. This is empty by default (deny-by-default: each aggregator is an explicit entry).

> The same audience discipline appears one more time, in the token exchange scenario. When Muster exchanges tokens with a target Dex, the target Dex is itself a validator of an incoming ID token—one whose audience is `muster`, not the target Dex. It accepts it because its connector is explicitly configured with which audience to expect on incoming tokens: Muster's client id at the source Dex. A Dex instance has no audience of its own as an issuer. Like every other validator in this document, it accepts exactly the audiences its configuration names, and nothing else.

Reviewer 3: Does it mean all Dex instances are configured with the audience of the central Dex?

Author: In practice all MCs, yeah. Each MC runs a connector (`giantswarm-simple-oidc`) configured to trust the central Dex as issuer and to expect Muster's client ID as the audience on incoming tokens. It's provisioned per management cluster together with the token-exchange credentials.

---
title: 'Authentication deep dive: from your MCP client to the Kubernetes API'
linkTitle: Authentication
description: A token-by-token walkthrough of how user identity flows from an MCP client through Muster and mcp-kubernetes to the Kubernetes API, covering login, validation, token forwarding, exchange, and impersonation.
diataxis_content_type: explanation
mermaid: true
weight: 45
menu:
  principal:
    parent: overview-ai-agents
    identifier: overview-ai-agents-authentication
owner:
  - https://github.com/orgs/giantswarm/teams/team-bumblebee
last_review_date: 2026-07-16
user_questions:
  - Which token does my MCP client send to Muster?
  - How does Muster validate tokens on every request?
  - How does my identity reach the Kubernetes API and its audit log?
  - How does Muster reach clusters that use a different Dex instance?
  - How do AI agents act on behalf of a human?
---

> You can configure several different MCP servers in Claude Code (or Cursor, and other MCP clients), and these MCP servers may or may not require authentication. For example, you may want to investigate the state of a Kubernetes cluster using `mcp-kubernetes`, which requires Kubernetes authentication. The challenge here is to keep the user identity through the whole communication, so at the very end, the requests to the Kubernetes API carry the user identity (and permissions).

Reviewer 2: What do you think about adding an introduction here that 1 points to the what muster is and oauth docs with links I think and 2 explain what muster authentication mechanisms are ? Becasue it's not really saying it supports oauth (token forwarding, token exchange or connecting to a remote authorization server right? I feel like this might simplify the content below more and make for a more shared understanding? Then for each flow we could add some diagrams to explain it best?

Reviewer 2: I'm not reviewing the content here but the format. What do you think about adding the main page to be the introduction and then have subpages like this:
- 1 page one how Muster authentication and oauth work with MCPs that might be a bit more high level
- another page that deep dives into how claude code access muster and the MCps behind it if necessary
- page on agent access to muster and OBO?

That way we can also introduce the agentgateway in the first one?

Reviewer 2: I'm sorry if this is a bit late in the process but the discussion frm yesterday shed some lights and I don't think we could have a shared understanding without splitting this into digestable chunks and without diagrams

Author: We do have diagrams. Please review using the local server to render the page properly and see all the content the way it's intented to be shown.

Author: I believe splitting the content in several smaller pages makes things actually harder to follow. In this page, we are telling the story about what happens when the user types `/mcp` in their claude code session. Splitting it in several pages makes it uncomfortable to follow and harder to consume IMO.

Also, this style was chosen purposely: instead of throwing definitions around (what's muster, what's oauth, etc), we explain why is necessary and what's happening.

  participant D as Dex<br/>(Muster's identity provider)
  participant IdP as Upstream IdP<br/>(GitHub / Azure)

  CC->>M: MCP request (no token)
  M-->>CC: 401 + WWW-Authenticate (metadata URL)
  CC->>B: open Muster /oauth/authorize<br/>client_id = CIMD URL, PKCE challenge
  B->>M: GET /oauth/authorize
  M-->>B: redirect to Dex /auth<br/>(Muster acting as Dex's client)
  B->>D: GET /auth
  D-->>B: redirect to IdP login page
  B->>IdP: user types password — only here
  IdP-->>B: redirect back to Dex callback + code
  B->>D: deliver code
  D->>IdP: exchange code (connector client_secret)
  IdP-->>D: user identity confirmed
  D-->>B: redirect to Muster /oauth/callback + new code
  B->>M: deliver code
  M->>D: exchange code at /token (Muster's Dex client_secret)
  D-->>M: Dex ID token + access token + refresh token
  M-->>B: redirect to localhost callback + final code
  B->>CC: deliver code
  CC->>M: exchange code + PKCE verifier
  M-->>CC: Muster access token (opaque)
  Note over M: stores map:<br/>Muster token → Dex token set
  Note over CC: Claude Code authenticates to Muster<br/>with the opaque token only
{{< /mermaid >}}
<!-- vale on -->

### OAuth clients

> In OAuth, the client is the application that wants tokens, and the user never types a password into it. A client redirects the user's browser to the Authorization Server (AS) and waits for the browser to come back with a one-time authorization code. It then exchanges that code for tokens over a back-channel call, identifying itself with its own credentials (`client_id` + `client_secret` for confidential clients like Muster). Public clients like Claude Code have no secret and use PKCE instead—see [below](#how-claude-code-becomes-a-client-without-pre-registration). Any application can be a client in one relationship and an Authorization Server in another. For example, Claude Code is a client of Muster, which is an AS to Claude Code but a client of Dex.

Reviewer 3: ```suggestion
In OAuth terms, the client is the application that requests tokens, and the user never types a password. A client redirects the user's browser to the Authorization Server (AS) and waits for the browser to return a one-time authorization code. It then exchanges that code for tokens over a back-channel call, identifying itself with its own credentials (`client_id` + `client_secret` for confidential clients like Muster). Public clients like Claude Code have no secret and use PKCE instead—see [below](#how-claude-code-becomes-a-client-without-pre-registration). Any application can be a client in one relationship and an AS in another. For example, Claude Code is a client of Muster, which is an AS to Claude Code but a client of Dex.
```

Author: Applied, with one small keep: I left "the user never types a password **into it**" intact. Dropping "into it" would make the sentence say the user never types a password at all — but they do, at the upstream identity provider; the point is that the *client* never sees it.

### How Claude Code becomes a client without pre-registration

The same team that deploys Muster also deploys Dex. They can feed a client_id/client_secret to both so that Muster (the client) can identify itself when talking to Dex, and Dex can check if the credentials are correct. But Muster's team can't know every MCP client that will ever connect.

So MCP clients identify themselves using Client ID Metadata Documents (CIMD). The `client_id` is an HTTPS URL, controlled by the client's vendor, hosting a JSON document that describes the client (its name, its redirect URIs). When Muster receives an authorization request with a URL-shaped `client_id`, it fetches that document and treats it as the client's registration. Nothing is stored ahead of time, and no secrets are exchanged between Muster's team and the client's vendor. This has two useful properties. First, client identity is anchored to domain ownership: only Anthropic can serve a document under an anthropic.com URL. So a malicious app can't credibly present itself as Claude Code. Second, every installation of the same client shares the one `client_id` (the URL), so Muster keeps no per-laptop registration state.

> (Muster also supports the older mechanism, Dynamic Client Registration (RFC 7591), as a fallback for clients that don't publish a metadata URL. The client POSTs its description to `/oauth/register` and receives a server-generated random `client_id`.)

Reviewer 3: ```suggestion
**Note**: Muster also supports the older mechanism, Dynamic Client Registration (RFC 7591), as a fallback for clients that don't publish a metadata URL. The client POSTs its description to `/oauth/register` and receives a server-generated random `client_id`.
```
We dont use parethesis in rest of docs

| Muster access token (opaque) | Muster | The user (via Muster's store) | Muster (resource binding recorded in the store) | Claude Code | Muster (checks 1–4 described earlier) |
| Muster refresh token | Muster | The user's session | Muster | Claude Code | Muster (rotation, 30-day session window) |
| Dex ID token (JWT) | Dex | The user (`sub`, `email`, `groups`) | `muster` (+ `dex-k8s-authenticator` via cross-client audience) | Muster, server-side | Downstream: mcp-kubernetes and the kube-apiserver, via Dex's JWKS |
| Dex access token | Dex | The user's Dex session | Dex itself (spent at `/userinfo`) | Muster, server-side | Dex, on every `/userinfo` call (check 3) |
| Dex refresh token | Dex | The user's Dex session | Dex itself | Muster, server-side | Dex, on refresh (check 2)—this is where upstream deprovisioning surfaces |

## What do these tokens actually look like

Muster's access token is opaque: a random string with no internal structure.

**Muster access token—opaque. What Claude Code sends on every request:**

> ```text

Reviewer 3: ```suggestion
Muster's access token is opaque: a random string with no internal structure. Look an example of what Claude Code sends on every request:

```text
```

Author: Applied to all three token examples (and the storage-record card too, for consistency) — the bold pseudo-headers are gone in favor of prose lead-ins. I adjusted the wording slightly ("Look **at** an example…", "look **at** what Claude Code redeems…") while keeping the structure you proposed.

Authorization: Bearer 3q7xKQm9vN1sYwZk8pR2tLc4bH6dJfA0uEgi5CnMoPw
```

There is nothing to decode. All its meaning lives in Muster's storage. Conceptually, the record that this string points at looks like this:

**Muster's storage record for that token (conceptual):**

```json
{
  "user": "user@giantswarm.io",
  "audience": "https://muster.example.gigantic.io/mcp",   // the RFC 8707 binding from check 4
  "expires_at": 1751638200,                               // 30 minutes after minting
  "dex_tokens": {
    "id_token": "eyJhbGciOi...",
    "access_token": "...",
    "refresh_token": "..."
  }
}
```

The Dex ID token is different: it's a JWT, three base64url-encoded segments separated by dots (header, payload, signature).

**Dex ID token—a signed JWT. Raw form, then the first two segments decoded (annotated):**

> 

Reviewer 3: ```suggestion
The Dex ID token is different: it's a JWT, three base64url-encoded segments separated by dots (header, payload, signature). Look at an example of an ID token. Raw form, then the first two segments decoded (annotated):

```

{ "alg": "RS256", "kid": "397a1bc4" }

// payload: the claims
{
  "iss": "https://dex.example.gigantic.io",    // who minted it        (Issuer check)
  "sub": "CiRqb3NlQGdp...",                    // opaque Dex user handle
  "aud": ["muster", "dex-k8s-authenticator"],  // who may consume it   (Audiences check)
  "azp": "muster",                             // which client actually requested it
  "email": "user@giantswarm.io",               // who it is about
  "groups": ["giantswarm:admins"],             // what kubernetes RBAC will evaluate
  "iat": 1751636400,                           // minted at
  "exp": 1751638200                            // expires at           (Expiration check)
}
```

Two things worth noticing. First, base64url is an encoding, not encryption: anyone holding this token can read every claim. The signature (the third segment) prevents *modification*, not reading. Second, each claim maps to one of the checks in ["How are Dex's ID tokens validated"](#how-are-dexs-id-tokens-validated) below. `iss`, `aud` and `exp` are the envelope. `email` and `groups` are the identity that will reach the Kubernetes RBAC evaluation and audit log.

> **Muster refresh token—opaque. What Claude Code redeems when the access token expires:**

Reviewer 3: ```suggestion
Now, look what Claude Code redeems when the access token expires:
```
